### DISCUSSION TITLE
SpotBit 2.0 - Intern Project

### DISCUSSION THREAD

**Discussion by dunchin7** [BlockchainCommons/Community#59] [Ideas]
# SpotBit 2.0

Spotbit is a portable API for Bitcoin price and candles. It can aggregate data from over 100 exchanges and serve them from a single URL or using Tor as an onion hidden service. It's extremely flexible: the user can decide which base currencies to use (USDT, USD, EUR etc), which exchanges to keep data for, and how much data to keep.

The Spotbit API, which is running right now has some architectural flaws as mentioned by @gorazdko. It is not efficient and the sequential use of threads for IO operations is very expensive. This generally leads to a deadlock state and server failure.
### Goal
* Rewrite the essential parts of SpotBit first in an efficient manner and then work towards features like:
    1) Anonymize over Tor
    2) Simple UI and Install Scripts

### Implementation:

* Use CCXT lib with asyncio to write concurrent code and solve the problem of threads.

* Which database should be used? InfluxDB vs Redis
Influx DB is a better option. (Many benefits over Redis)
(Can be decided after completing initial milestones.)

### Milestones

* Discuss potential ideas and integrations with Spotbit and Develop a roadmap for Spotbit 2.0.

* **Implementation (30-07-2021)**
    

   - [ ] Understand the working of CCXT with asyncio 

    
  - [ ] Develop a REST API for Spotbit(like what we have now) without DB and using asyncio.

    

  - [ ] Provide self-sovereignty with commands like- now/currency/exchange, /now, /configure.

    

   - [ ] Test the application and fix if any bugs.

Further goals would be to scope 
* Tor-based implementation and adding DB for supporting historical prices.
* Basic UI and Bitcoin fee estimation model.

**Comment by nochiel** [comment]
It might be worth considering using a user-friendly asyncio library if you need to do heavy lifting? Something like [trio](https://trio.readthedocs.io/en/stable/) ?
